tpmutil/run_windows.go

```diff
@@ -0,0 +1,172 @@
+	0x80284005: "The specified output buffer is too small.",
+	0x80284006: "An error occurred while communicating with the TPM.",
+	0x80284007: "A context parameter that is not valid was passed when attempting to create a TBS context.",
+	0x80284008: "The TBS service is not running and could not be started.",
+	0x80284009: "A new context could not be created because there are too many open contexts.",
+	0x8028400A: "A new virtual resource could not be created because there are too many open virtual resources.",
+	0x8028400B: "The TBS service has been started but is not yet running.",
+	0x8028400C: "The physical presence interface is not supported.",
+	0x8028400D: "The command was canceled.",
+	0x8028400E: "The input or output buffer is too large.",
+	0x8028400F: "A compatible Trusted Platform Module (TPM) Security Device cannot be found on this computer.",
+	0x80284010: "The TBS service has been disabled.",
+	0x80284011: "The TBS event log is not available.",
+	0x80284012: "The caller does not have the appropriate rights to perform the requested operation.",
+	0x80284013: "The TPM provisioning action is not allowed by the specified flags.",
+	0x80284014: "The Physical Presence Interface of this firmware does not support the requested method.",
+	0x80284015: "The requested TPM OwnerAuth value was not found.",
+}
+
+func tbsError(err uintptr) error {
+	if err == 0 {
+		return nil
+	}
+	if description, ok := errMap[err]; ok {
+		return fmt.Errorf("TBS Error: %s", description)
+	}
+	return fmt.Errorf("TBS Error: %v", err)
+}
+
+// winTPMBuffer is a ReadWriteCloser to access the TPM in Windows.
+type winTPMBuffer struct {
+	hContext  uintptr
+	outBuffer []byte
+	priority  CommandPriority
+}
+
+// Executes the TPM command specified by inBuff, returning the number of bytes in the command
+// and any error code returned by executing the TPM command. Command response can be read by calling
+// Read().
+func (rwc *winTPMBuffer) Write(commandBuffer []byte) (int, error) {
+	// TPM spec defines longest possible response to be maxTPMResponse
+	outBufferLen := maxTPMResponse
+	rwc.outBuffer = make([]byte, outBufferLen)
+
+	// TBS_RESULT Tbsip_Submit_Command(
+	//   _In_          TBS_HCONTEXT         hContext,
+	//   _In_          TBS_COMMAND_LOCALITY Locality,
+	//   _In_          TBS_COMMAND_PRIORITY Priority,
+	//   _In_    const PCBYTE               *pabCommand,
+	//   _In_          UINT32               cbCommand,
+	//   _Out_         PBYTE                *pabResult,
+	//   _Inout_       UINT32               *pcbOutput
+	// );
+	errResp, _, _ := tbsSubmitCommand.Call(
+		rwc.hContext,
+		tbsCommandLocalityZero, // Windows currently only supports TBS_COMMAND_LOCALITY_ZERO
+		uintptr(rwc.priority),
+		uintptr(unsafe.Pointer(&(commandBuffer[0]))),
+		uintptr(len(commandBuffer)),
+		uintptr(unsafe.Pointer(&(rwc.outBuffer[0]))),
+		uintptr(unsafe.Pointer(&outBufferLen)),
+	)
+
+	// shrink outBuffer so it is length of response
```

Comment:
Please start comments with a capital letter and end with punctuation.

Reply:
Done